vllm/utils/deep_gemm.py

```diff
@@ -0,0 +1,135 @@
+    _per_token_cast_impl: Callable[..., Any] | None = None
+    _per_block_cast_impl: Callable[..., Any] | None = None
+else:
+    _dg = importlib.import_module("deep_gemm")  # type: ignore
+
+    _fp8_gemm_nt_impl = _resolve_symbol(
+        _dg,
+        "fp8_gemm_nt",
+        "gemm_fp8_fp8_bf16_nt",
+    )
+    _grouped_impl = _resolve_symbol(
+        _dg,
+        "m_grouped_fp8_gemm_nt_contiguous",
+        "m_grouped_gemm_fp8_fp8_bf16_nt_contiguous",
+    )
+    _grouped_masked_impl = _resolve_symbol(
+        _dg,
+        "fp8_m_grouped_gemm_nt_masked",
+        "m_grouped_gemm_fp8_fp8_bf16_nt_masked",
+    )
+
+    # Try to get per_token_cast_to_fp8 from DeepGEMM math utils.
+    try:
+        _math_mod = importlib.import_module(
+            "deep_gemm.utils.math")  # type: ignore
+        _per_token_cast_impl = getattr(_math_mod, "per_token_cast_to_fp8",
+                                       None)
+        _per_block_cast_impl = getattr(_math_mod, "per_block_cast_to_fp8",
+                                       None)
+    except ModuleNotFoundError:
+        _per_token_cast_impl = None
+        _per_block_cast_impl = None
+
+
+def fp8_gemm_nt(*args, **kwargs):
+    if _fp8_gemm_nt_impl is None:
+        return _missing(*args, **kwargs)
+    return _fp8_gemm_nt_impl(*args, **kwargs)
+
+
+def m_grouped_fp8_gemm_nt_contiguous(*args, **kwargs):
+    if _grouped_impl is None:
+        return _missing(*args, **kwargs)
+    return _grouped_impl(*args, **kwargs)
+
+
+def fp8_m_grouped_gemm_nt_masked(*args, **kwargs):
+    if _grouped_masked_impl is None:
+        return _missing(*args, **kwargs)
+    return _grouped_masked_impl(*args, **kwargs)
+
+
+def per_token_group_cast_to_fp8(x, group_size, *args, **kwargs):
+    """Wrapper for token-wise FP8 quantisation.
+
+    • If DeepGEMM provides ``per_token_cast_to_fp8`` (new API), use it.
+    • Otherwise, fall back to vLLM's ``per_token_group_quant_fp8``
+    """
+    if _per_token_cast_impl is not None:
+        assert group_size == 128, "group_size must be 128 for deepgemm"
+        return _per_token_cast_impl(x)
+
+    from vllm.model_executor.layers.quantization.utils.fp8_utils import (
+        per_token_group_quant_fp8 as _ptg)
+    return _ptg(x, group_size, *args, **kwargs)
+
+
+def per_block_cast_to_fp8(x, *args, **kwargs):
+    if _per_block_cast_impl is not None:
+        return _per_block_cast_impl(x)
+    # TODO: refactor the `per_block_cast_to_fp8` from tests to vllm utils
+    from tests.kernels.quant_utils import per_block_cast_to_fp8 as _pbcf
+    return _pbcf(x, *args, **kwargs)
+
+
+def calc_diff(x: torch.Tensor, y: torch.Tensor):
+    x, y = x.double(), y.double()
+    denominator = (x * x + y * y).sum()
+    sim = 2 * (x * y).sum() / denominator
+    return 1 - sim
+
+
+@functools.cache
+def is_blackwell_deep_gemm_used():
+    import vllm.envs as envs
+    return envs.VLLM_USE_DEEP_GEMM and has_deep_gemm(
+    ) and _per_block_cast_impl is not None
```

Comment:
Should we also check that `current_platform.is_device_capability(100)`?

Reply:
Certainly, thanks! fixed